play-ahc-ws-standalone/src/main/scala/play/api/libs/ws/ahc/StandaloneAhcWSClient.scala

```diff
@@ -163,5 +156,106 @@ object StandaloneAhcWSClient {
     new SystemConfiguration(loggerFactory).configure(config.wsClientConfig.ssl)
     wsClient
   }
+
+  /**
+   * Ensures:
+   * 1. [[StandaloneWSRequest.url]] path is encoded.
+   * 2. Any query params present in the URL are moved to [[StandaloneWSRequest.queryString]].
+   */
+  @throws[IllegalArgumentException]("if the url is unrepairable")
+  private[ahc] def normalize(req: StandaloneWSRequest): StandaloneWSRequest = {
+    try {
+      // Recover from https://github.com/AsyncHttpClient/async-http-client/issues/1149
+      Uri.create(req.url)
+      if (req.uri.getQuery == null) {
```

Comment:
If `req.uri.getQuery == null` then you want to try `Uri.create(req.url)` -- otherwise you'll throw and catch an exception but will be calling `repair()` in any case.  If you can avoid throwing the exception it'll speed the happy path.

Reply:
I need some clarification to understand this comment.

> If `req.uri.getQuery == null` then you want to try `Uri.create(req.url)--`

`Uri.create(req.url)` is called on the line above, 169.

> otherwise you'll throw and catch an exception

> If you can avoid throwing the exception it'll speed the happy path.

Which exception? The one from `ahc.Uri` or `java.net.URI`? `ahc.Uri` is tolerant and should only throw if critical parts are missing, e.g. https://github.com/AsyncHttpClient/async-http-client/issues/1149. `java.net.URI` is stricter and I'd expect it throw if the query params are not encoded. I wouldn't consider either to be on the happy path since they're both avoidable by the caller.

Reply:
So something like

```
 @throws[IllegalArgumentException]("if the url is unrepairable")
  private[ahc] def normalize(req: StandaloneWSRequest): StandaloneWSRequest = {
    if (req.uri.getQuery == null) {
      req
    } else {
        try {
          Uri.create(req.url)
          repair(req)
        } catch {
          case NonFatal(_) =>
            // URI parsing error
            repair(req)
        }
    }
}
```